backend/src/openai.ts

```diff
@@ -218,50 +218,55 @@ async function chatGptCallFunction(
 	};
 }
 
-async function chatGptChatCompletion(
-	chatResponse: ChatResponse,
-	chatHistory: ChatHistoryMessage[],
+function setSystemRoleInChatHistory(
+	currentLevel: LEVEL_NAMES,
 	defences: Defence[],
-	chatModel: ChatModel,
-	openai: OpenAI,
-	// default to sandbox
-	// eslint-disable-next-line @typescript-eslint/no-unused-vars
-	currentLevel: LEVEL_NAMES = LEVEL_NAMES.SANDBOX
+	chatHistory: ChatHistoryMessage[]
 ) {
-	// check if we need to set a system role
-	// system role is always active on levels
-	if (
-		currentLevel !== LEVEL_NAMES.SANDBOX ||
-		isDefenceActive(DEFENCE_ID.SYSTEM_ROLE, defences)
-	) {
+	const systemRoleNeededInChatHistory =
+		currentLevel === LEVEL_NAMES.SANDBOX
+			? isDefenceActive(DEFENCE_ID.SYSTEM_ROLE, defences)
+			: true;
+
+	if (systemRoleNeededInChatHistory) {
 		const completionConfig: ChatCompletionSystemMessageParam = {
 			role: 'system',
 			content: getSystemRole(defences, currentLevel),
 		};
 
-		// check to see if there's already a system role
-		const systemRole = chatHistory.find(
+		const existingSystemRole = chatHistory.find(
 			(message) => message.completion?.role === 'system'
 		);
-		if (!systemRole) {
-			// add the system role to the start of the chat history
-			chatHistory.unshift({
-				completion: completionConfig,
-				chatMessageType: CHAT_MESSAGE_TYPE.SYSTEM,
-			});
+		if (!existingSystemRole) {
+			return [
+				{
+					completion: completionConfig,
+					chatMessageType: CHAT_MESSAGE_TYPE.SYSTEM,
+				},
+				...chatHistory,
+			];
 		} else {
-			// replace with the latest system role
-			systemRole.completion = completionConfig;
+			return chatHistory.map((message) => {
+				if (message.completion?.role === 'system') {
+					return { ...existingSystemRole, completion: completionConfig };
+				} else {
+					return message;
+				}
+			});
 		}
 	} else {
-		// remove the system role from the chat history
-		while (
-			chatHistory.length > 0 &&
-			chatHistory[0].completion?.role === 'system'
-		) {
-			chatHistory.shift();
-		}
+		return chatHistory.filter(
+			(message) => message.completion?.role !== 'system'
+		);
```

Comment:
I was going to say this would be less efficient, but a quick google suggests `shift` to be O(N) 🤷 
So overall this reads a lot nicer, AND is likely to be just as efficient, if not more so. Nice one!

Reply:
I'm surprised shift() is O(N)! I guess it moves all the values about in memory so there isn't a dead bit of memory where the first element used to be